src/limb.rs

```diff
@@ -0,0 +1,271 @@
+        for i in 0..(LIMB_BITS - 1) {
+            let bit = 1 << (LIMB_BITS - 1 - i);
+            if self.most_significant_limb() & bit > 0 {
+                return i;
+            }
+        }
+
+        unreachable!();
+    }
+
+    fn is_valid(&self) -> bool {
+        if self.limbs.len() == 0 {
+            false
+        } else if self.most_significant_limb() == 0 {
+            false
+        } else if self.limbs.len() == 1 && self.most_significant_limb() == 1 {
+            false
+        } else {
+            true
+        }
+    }
+
+    /// Are these little-endian limbs within the range? Checks in constant time.
+    pub fn are_limbs_within(&self, limbs: &[Limb]) -> bool {
+        assert_eq!(self.limbs.len(), limbs.len());
+
+        let is_zero = limbs_are_zero_constant_time(limbs);
+        let is_lt_n = limbs_less_than_limbs_constant_time(limbs, self.limbs);
+
+        is_lt_n && !is_zero
+    }
+
+    // The code in this function, including the comments, was largely copied
+    // from `ec::suite_b::private_key::generate_private_key`, with adaptations.
+    //
+    // TODO Merge the two implementations.
+    //
+    // NSA Guide Appendix B.2: "Key Pair Generation by Testing Candidates".
+    //
+    // TODO: The NSA guide also suggests, in appendix B.1, another mechanism
+    // that would avoid the need to use `rng.fill()` more than once. It works
+    // by generating an extra 64 bits of random bytes and then reducing the
+    // output (mod n). Supposedly, this removes enough of the bias towards
+    // small values from the modular reduction, but it isn't obvious that it is
+    // sufficient. TODO: Figure out what we can do to mitigate the bias issue
+    // and switch to the other mechanism.
+    //
+    /// Chooses a positive integer within the range and stores it into `dest`
+    /// encoded as little-endian limbs.
+    ///
+    /// This function is intended to be suitable for generating private keys.
+    pub fn sample_into_limbs(&self, dest: &mut [Limb], rng: &rand::SecureRandom)
+                             -> Result<(), ()> {
+
+        // NSA Guide Steps 1, 2, and 3.
+        assert_eq!(self.limbs.len(), dest.len());
+        debug_assert!(self.is_valid());
+
+        // XXX: The value 100 was chosen to match OpenSSL due to uncertainty of
+        // what specific value would be better, but it seems bad to try 100 times.
+        for _ in 0..100 {
+
+            // NSA Guide Step 4.
+            //
+            // The requirement that the random number generator has the requested
+            // security strength is delegated to `rng`.
+            {
+                let mut dest_as_bytes = limbs_as_bytes_mut(dest);
+                try!(rng.fill(&mut dest_as_bytes));
+            }
+
+            let bits_to_mask = self.leading_zero_bits();
+            if bits_to_mask > 0 {
+                let keep: Limb = (1 << (LIMB_BITS - bits_to_mask)) - 1;
+                dest[self.limbs.len() - 1] &= keep;
+            }
```

Comment:
Definitely I don't want to do the `cfg(attr)` thing. I think, at a minimum, we should pull the calculation of the mask out of the loop so that it is clearer that it doesn't change from one interation to another.